### DISCUSSION TITLE
unit for flows related to LID [FLOW_UNITS LPS]

### DISCUSSION THREAD

**Discussion by vww-wrc** [pyswmm/pyswmm#340] [Q&A] [answered]
Hi, I am trying to understand the inflow/outflow of LID over time using pyswmm and I am confused about the units of data. My SWMM inp file's unit is LPS, does that mean the flow rates in/out of LID are in liters/second, or is it meters/sec? Can I multiply the rate with timestep in seconds to get volume? 

Please see below the type of data I am trying to obtain and understand, with questions on each data following # sign. 

with Simulation("template.inp") as sim:

     lid_on_sub = LidGroups(sim)["SC_WithLID"]
for step in sim:

     print(lid_on_sub[0].surface.inflow)    
      # From SWMM source code lidproc.c page line 122 it says this is: "precip. + runon to LID unit (ft/s)". But is it m/s in LPS? 
      # If my reporting/routing time step is 1min, can I just multiple this rate with 60 to get runoff volume? 
     print(lid_on_sub[0].soil.percolation)  # is this saturated hydraulic conductivity in soil layer? mm/hr? 
     print(lid_on_sub[0].storage.drain) # underdrain flow rate in m/s? 
     print(lid_on_sub[0].surface.depth) # is this in mm? 
     print(lid_on_sub[0].surface.outflow) # m/s? 
     print(lid_on_sub[0].storage.depth) # mm? 
     print(lid_on_sub[0].storage.exfiltration) #mm/hr?  

SWMM source code lidproc.c page:  [https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/develop/src/solver/lidproc.c](url)

Thank you!

**Comment by bemcdonnell** [comment]
@vww-wrc until @jennwuu has a minute to respond on the particular units, here is the outbound function from SWMM to PySWMM in the toolkit.  

https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/toolkit.c#L2675-L2772

The unit conversion takes place for each result.

The unit conversion values are here:

https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/swmm5.c#L124-L144

**Reply by vww-wrc** [reply]
Thank you very much! @bemcdonnell

**Comment by jennwuu** [comment, marked as answer]
Hi @vww-wrc,

Good questions!

The internal units used by SWMM engine is ft, ft/s, etc: https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/lid.h#L72-L135

In the API, the ft unit is converted to SI units (like mm or mm/hr): 
https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/develop/src/solver/toolkit.c#L2675-L2772
https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/toolkit.c#L2626-L2673
https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/develop/src/solver/toolkit.c#L2575-L2624

I will try to address your questions with reference to code in pyswmm and OWA/SWMM engine.

Questions:
```
print(lid_on_sub[0].surface.inflow)    
# From SWMM source code lidproc.c page line 122 it says this is: "precip. + runon to LID unit (ft/s)". But is it m/s in LPS? 
# If my reporting/routing time step is 1min, can I just multiple this rate with 60 to get runoff volume? 
``` 

Pyswmm: 
https://github.com/OpenWaterAnalytics/pyswmm/blob/a73409b2a9ebf4131ba42fc4c03a69453fc29d9a/pyswmm/lidunits.py#L43-L54
 
OWA/SWMM: 
- https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/toolkit.c#L2712-L2713
- https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/develop/src/solver/lid.h#L157

The internal unit is ft, so after API conversion, the data provided is in mm via UCF(RAINDEPTH). 

```print(lid_on_sub[0].soil.percolation)  # is this saturated hydraulic conductivity in soil layer? mm/hr? ```

PySWMM: 
https://github.com/OpenWaterAnalytics/pyswmm/blob/master/pyswmm/lidunits.py#L274-L285

https://github.com/OpenWaterAnalytics/pyswmm/blob/master/pyswmm/toolkitapi.py#L199

OWA/SWMM:
- https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/toolkit.c#L2752-L2753
- https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/lid.h#L175

The internal unit is ft/s, so after API conversion, the data provided is in mm/hr via UCF(RAINFALL). 

```print(lid_on_sub[0].storage.drain) # underdrain flow rate in m/s? ```

PySWMM: 
https://github.com/OpenWaterAnalytics/pyswmm/blob/a73409b2a9ebf4131ba42fc4c03a69453fc29d9a/pyswmm/lidunits.py#L217-L227

https://github.com/OpenWaterAnalytics/pyswmm/blob/master/pyswmm/toolkitapi.py#L203

OWA/SWMM:
- https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/toolkit.c#L2760-L2761
- https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/develop/src/solver/lid.h#L179

The internal unit is ft/s, so after API conversion, the data provided is in mm/hr via UCF(RAINFALL). 

```print(lid_on_sub[0].surface.depth) # is this in mm? ```

PySWMM: 
https://github.com/OpenWaterAnalytics/pyswmm/blob/a73409b2a9ebf4131ba42fc4c03a69453fc29d9a/pyswmm/lidunits.py#L31-L41

https://github.com/OpenWaterAnalytics/pyswmm/blob/a73409b2a9ebf4131ba42fc4c03a69453fc29d9a/pyswmm/toolkitapi.py#L181

OWA/SWMM:
- https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/toolkit.c#L2720-L2721
- https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/lid.h#L207

The internal unit is ft, so after API conversion, the data provided is in mm via UCF(RAINDEPTH). 

```print(lid_on_sub[0].surface.outflow) # m/s? ```
PySWMM: 
https://github.com/OpenWaterAnalytics/pyswmm/blob/master/pyswmm/lidunits.py#L79-L89

https://github.com/OpenWaterAnalytics/pyswmm/blob/a73409b2a9ebf4131ba42fc4c03a69453fc29d9a/pyswmm/toolkitapi.py#L195

OWA/SWMM:
- https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/develop/src/solver/toolkit.c#L2744-L2745
- https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/lid.h#L171

The internal unit is ft/s , so after API conversion, the data provided is in mm/hr via UCF(RAINFALL).
 
```print(lid_on_sub[0].storage.depth) # mm? ```
PySWMM: 
https://github.com/OpenWaterAnalytics/pyswmm/blob/a73409b2a9ebf4131ba42fc4c03a69453fc29d9a/pyswmm/lidunits.py#L169-L179

https://github.com/OpenWaterAnalytics/pyswmm/blob/a73409b2a9ebf4131ba42fc4c03a69453fc29d9a/pyswmm/toolkitapi.py#L184

OWA/SWMM:
- https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/develop/src/solver/toolkit.c#L2726-L2727
- https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/lid.h#L210

The internal unit is ft, so after API conversion, the data provided is in mm via UCF(RAINDEPTH). 

```print(lid_on_sub[0].storage.exfiltration) #mm/hr? ```
PySWMM: 
https://github.com/OpenWaterAnalytics/pyswmm/blob/master/pyswmm/lidunits.py#L193-L203

https://github.com/OpenWaterAnalytics/pyswmm/blob/a73409b2a9ebf4131ba42fc4c03a69453fc29d9a/pyswmm/toolkitapi.py#L201

OWA/SWMM:
- https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/toolkit.c#L2756-L2757

The internal unit is ft/s, so after API conversion, the data provided is in mm/hr via UCF(RAINFALL). 

I think we should add units in the documentation based on your questions. Feel free to submit a PR to add units based on your understanding of LID API, this will be really helpful for the community 🦥

**Reply by vww-wrc** [reply]
Thank you very much @jennwuu for the detailed answer! This is very helpful and I will submite a PR soon. Just to double check - I think the unit for LID surface.inflow should be mm/hr as it's not the same as surfFlow in water balance whose unit is mm (volume), correct? 

![fig](https://user-images.githubusercontent.com/83201399/171398055-9be3a13a-1af9-4507-8927-38ca6b8cb843.png)

Thanks,
Vivian

**Comment by jennwuu** [comment]
Hi @vww-wrc 

**lid_unit.surface.inflow** 

PySWMM:
https://github.com/OpenWaterAnalytics/pyswmm/blob/a73409b2a9ebf4131ba42fc4c03a69453fc29d9a/pyswmm/lidunits.py#L43-L54
https://github.com/OpenWaterAnalytics/pyswmm/blob/a73409b2a9ebf4131ba42fc4c03a69453fc29d9a/pyswmm/toolkitapi.py#L192

OWA/SWMM
https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/toolkit.c#L2738-L2739

The internal unit is ft/s so after API conversion, the data provided is in mm/hr via UCF(RAINFALL).

**lid_unit.WaterBalance.surface_flow**

PySWMM
https://github.com/OpenWaterAnalytics/pyswmm/blob/f6edd28f97a52b67018981f64f311d7a005e23e7/pyswmm/lidgroups.py#L227
https://github.com/OpenWaterAnalytics/pyswmm/blob/master/pyswmm/lidunits.py#L343-L353
https://github.com/OpenWaterAnalytics/pyswmm/blob/a73409b2a9ebf4131ba42fc4c03a69453fc29d9a/pyswmm/toolkitapi.py#L177

OWA/SWMM
https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/6d6b5272250acaa3939db9e265f0e593dc96e43e/src/solver/toolkit.c#L2712-L2713

The internal unit is ft, so after API conversion, the data provided is in mm via UCF(RAINDEPTH).

Can't wait for the PR! Thanks!!
